libs/s25main/mapGenerator/Algorithms.h

```diff
@@ -225,47 +225,36 @@ namespace rttr { namespace mapGenerator {
     template<typename T_Value>
     NodeMapBase<unsigned> Distances(const MapExtent& size, T_Value&& evaluator)
     {
-        const unsigned maximumDistance = size.x * size.y;
-
-        std::queue<MapPoint> queue;
-        NodeMapBase<unsigned> distances;
-        distances.Resize(size, maximumDistance);
-
+        std::vector<MapPoint> flaggedPoints;
         RTTR_FOREACH_PT(MapPoint, size)
         {
             if(evaluator(pt))
             {
-                distances[pt] = 0;
-                queue.push(pt);
+                flaggedPoints.push_back(pt);
             }
         }
-
-        UpdateDistances(distances, queue);
-
-        return distances;
+        return DistancesTo(flaggedPoints, size);
     }
 
     /**
      * Counts the number of values within the specified range.
      *
-     * @param values map of comparable values
+     * @param nodes map of comparable values
      * @param minimum minimum value to consider
      * @param maximum maximum value to consider
      *
      * @returns number of values between the specified minimum and maximum values.
      */
     template<typename T>
-    unsigned Count(const NodeMapBase<T>& values, T minimum, T maximum)
+    unsigned Count(const NodeMapBase<T>& nodes, T minimum, T maximum)
     {
-        unsigned valuesInRange = 0;
-        RTTR_FOREACH_PT(MapPoint, values.GetSize())
-        {
-            if(values[pt] >= minimum && values[pt] <= maximum)
-            {
-                valuesInRange++;
-            }
-        }
-        return valuesInRange;
+        std::function<unsigned(const std::vector<T>&)> countBetween = [minimum, maximum](const std::vector<T>& values) {
```

Comment:
```suggestion
        auto countBetween = [minimum, maximum](const std::vector<T>& values) {
```
https://www.reddit.com/r/cpp/comments/kzvjgn/c_performance_trap_2_unnecessary_stdfunction/
Keep this a lambda and then we can change the `nodes.Map` later to also take a lambda. `std::function` has some heavy overhead: https://www.reddit.com/r/cpp/comments/kzvjgn/c_performance_trap_2_unnecessary_stdfunction/ (found recently). Note that the rest does not need to change but it avoids changing this again

Reply:
When I try to pass it in as a lambda, I get:
`No matching member function for call to 'Map'`

I guess it doesn't like the template return type `T_Result` of `Map` in combination with a lambda?

Reply:
The idea of the `NodeMapBase.Map` function was to have a more efficient way to perform operations on the stored nodes by operating directly on the `std::vector<Node>` instead of iterating over all `MapPoint`s. Also, that way I could use standard functions like `count_if` which makes the code more readable. The obvious disadvantage that I'm exposing the underlying storage of the `NodeMapBase` (but only as `const ref`). 

I also thought about splitting up the `Map` function into more concrete ones, like `unsigned Count(lambda_describing_which_values_to_count)`. Maybe there's a nicer way to achieve this?

Reply:
Ah yes, good point, sorry I missed that.
The idea is good, but maybe for this a getter is better or even adding (proxy) begin/end functions so you could even do `for(const auto& node: map)`, but I guess `for(const auto& node: map.getNodes())` is good enough (side note: I want to get to having method names lowerCase for consistency and distinguishing from types)
The reason is that for `map` I expected that the lambda is called for each node, which is the usual meaning of "map". We could add `accumulate(lambda, startVal)` and a "real" `map(lambda)`, but that depends on what we need. Judging from the current usages I'd say `getNodes` or `begin/end` is the best and removing the `Map` function. The latter does not fully expose the storage class, so maybe that is better and it allows exposing non-const iterators (exposing a non-const vector is impossible as changing its size would break the class)

And maybe add an inverse `getIdx`: `MapPoint getPosition(unsigned index)`(name?) (you used that functionality in one of the Map-calls)

What do you say? Would you be up for those changes? (Maybe even in this PR as it looks related)

> I guess it doesn't like the template return type T_Result of Map in combination with a lambda?

Yeah it can't deduce it. So `map.Map<unsigned>(..)` would need to be used or something like `template<typename T> auto Map(T&& func) const;`

Reply:
Yes, definitely like your suggestions. I'll take a look at this after work!